Assets/MixedRealityToolkit/MSBuild/Scripts/TemplateFiles.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#if UNITY_EDITOR
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using UnityEditor;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.MSBuild
+{
+    /// <summary>
+    /// A helper class to manage (and locate) all the templates.
+    /// </summary>
+    public class TemplateFiles
+    {
+        private const string TemplateFilesFolderName = "MSBuildTemplates";
+        private const string MSBuildSolutionTemplateName = "SolutionTemplate.sln";
+        private const string SDKProjectFileTemplateName = "SDKProjectTemplate.csproj";
+        private const string PlatformPropsTemplateName = "Platform.Configuration.Template.props";
+        private const string EditorPropsTemplateName = "Editor.InEditor.Template.props";
+        private const string SpecifcPlatformPropsTemplateRegex = @"[a-zA-Z]+\.[a-zA-Z]+\.Template\.props";
+
+        private static TemplateFiles instance;
+
+        /// <summary>
+        /// Gets the singleton instance (created on demand) of this class.
+        /// </summary>
+        public static TemplateFiles Instance => instance ?? (instance = new TemplateFiles());
+
+        private string generatedCSMetaTemplateText;
+
+        /// <summary>
+        /// Gets the MSBuild Solution file (.sln) template path.
+        /// </summary>
+        public string MSBuildSolutionTemplatePath { get; }
+
+        /// <summary>
+        /// Gets the MSBuild C# SDK Project file (.csproj) template path.
+        /// </summary>
+        public string SDKProjectFileTemplatePath { get; }
+
+        /// <summary>
+        /// Gets the MSBuild Platform Props file (.props) template path.
+        /// </summary>
+        public string PlatformPropsTemplatePath { get; }
+
+        /// <summary>
+        /// Gets a list of specialized platform templates.
+        /// </summary>
+        public IReadOnlyDictionary<string, string> PlatformTemplates { get; }
+
+        /// <summary>
+        /// Gets a list of all other files included among the templates.
+        /// </summary>
+        public IReadOnlyList<string> OtherFiles { get; }
+
+        private TemplateFiles()
+        {
+            string[] templateFolders = AssetDatabase.FindAssets(TemplateFilesFolderName);
+            Utilities.GetPathsFromGuidsInPlace(templateFolders);
+
+            if (templateFolders.Length == 0)
+            {
+                Debug.LogError($"Templates folder '{TemplateFilesFolderName}' not found.");
+            }
+            else if (templateFolders.Length > 1)
+            {
+                Debug.LogWarning($"Stange, more than one directory exists for template files:\n {string.Join("\n", templateFolders)}");
+            }
```

Comment:
Should either of these end up throwing as well (i.e. if we can't get any template folders, is that actually super bad for this, or is this... an acceptable situation to be in?)

Reply:
Yeah, i don't know. I am relying on the consumer throwing when it doesn't find the template.

---
In reply to: [302782608](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5215#discussion_r302782608) [](ancestors = 302782608)